### ISSUE TITLE
.rclone.conf becoming owned by a different user with different permissions

### ISSUE THREAD

**Issue by calisro** [ncw/rclone#1467]
Somewhere between beta.v1.36-49-g5135ff7β  and v1.36-158-ga5cfdfd2β something has changed in the way the .rclone.conf file is being saved to disk.  In my setup I have rclone mounts running as root.  I also use rclone scripted as my non-root user to do some activities.

I have found that the .rclone.conf file is getting replaced (probably by the mounts running as root) and changing the ownership to root.  It shoudn't be root.  

If I change it back to my user, after some time it changes back to root.  This breaks other processes I have running on the box that require access to that file as it is unreadable.  

I would guess it is this commit that introduced the problem.

https://github.com/ncw/rclone/commit/9cede6b372002a739cca8d5db589c61c6d229bba

https://github.com/ncw/rclone/issues/1287

It looks like a temp config file is being created to mitigate some risk on blowing up the config file.  When that is done, it should be creating that new config with exactly the same permissions, ownership, and even ACLs.

**Comment by calisro** [comment]
Also looking at that commit, I believe this can cause problems if multiple rclone processes are running. 	
If the config file is renamed to an old one first then any other process trying to read that file at that time will fail.

```
+ if err = os.Rename(ConfigPath, ConfigPath+".old"); err != nil && !os.IsNotExist(err) {
 +		log.Fatalf("Failed to move previous config to backup location: %v", err)
 +	}
 +	if err = os.Rename(f.Name(), ConfigPath); err != nil {
 +		log.Fatalf("Failed to move newly written config from %s to final location: %v", f.Name(), err)
 +	}
 +	if err := os.Remove(ConfigPath + ".old"); err != nil && !os.IsNotExist(err) {
 +		Errorf(nil, "Failed to remove backup config file: %v", err)
 +	}
```

**Comment by breunigs** [comment]
Regarding the permissions: rclone has always set those to `0600`. I guess it previously worked for you because rclone would continue even if it couldn't set the permissions. For instance, if a file is owned by `root:root` with `0777`, any user can write to it, but not change permissions. I am not sure what's a good approach here, as `0600` is a sensible permission to have, considering the config file contains credentials.

Regarding the user: It can be kept, if and only if the current user is privileged enough. This would work for root+user, but probably not for root+user1+user2.

Regarding the race condition: I would use a lock file instead, which should solve everything nicely, both reading and writing.

To sum things up:
- the race condition is an obvious flaw, that is straightforward to fix. Good catch!
- I don't see any decent solution for the owner/permission issue yet. As a workaround you can copy the file to every user that needs it when you do manual changes (and adjust permissions).

**Comment by calisro** [comment]
For hte permissions, instead of recreating the files, if you simply change the contents, the permissions do not get changed.  For example in Linux If I 'cat' new data into an existing file the permissions on that file and the ownership remains the same.  Perhaps instead of moving/renaming the main file you change the contents:

read config
write config.new
delete config.old if it exists
COPY config config.old
go equivalent of 'cat' into the existing config.

That would solve all the permissions problems and ownership.

**Comment by breunigs** [comment]
`cat`, `copy` and "write to file" are essentially the same: read some bytes, write some bytes, repeat until done. `copy`, or at least the tool `cp` also sets permissions on the new file. Writing bytes is unfortunately not atomic, so we might need multiple system calls to actually write a file. If the disk is full, we might end up with a half-written and thus broken file.

If I am not mistaken, your last suggestion would re-open #1287 when `cat` fails due to the disk being full. That's why the patch uses `rename` in the first place: this action is atomic and thus doesn't leave files in an undetermined state – it either works or doesn't.

**Comment by calisro** [comment]
Hmm.  Agreed but there has to be a better way to protect the file on a full disk without breaking changed permissions/ownership of the config and existing use cases.  My thought process was if the disk is full then the last write would fail and if that were the case, rclone should fail anyway and remove the half written file if it exists.  On the next run, rclone should be able to determine that it failed previously by detecting the presence of the old file and adjusting itself by using that backup file.

read config
write config.new
delete config.old if it exists
COPY config config.old  
<===== Disk FULL
if write to new config FAILS then abort and attempt to remove partial file.
Next time rclone starts, and there is no config and there is a 'old' config then it should know it failed and fix the config by using the backup 'old' copy.

**Comment by calisro** [comment]
read config
write config.new
delete config.old if it exists
COPY config config.old
<===== Disk FULL
if write to new config.new to config FAILS then abort and attempt to rewrite to that file from the old file.

**Comment by calisro** [comment]
This is sort of what I was thinking to prevent the loss of permissions/owners while still protecting the config.

if  config.old then 
    assume previous write failure
    write config.old config
end if

read config
obtain lock
write config.new
delete config.old if it exists
copy config config.old
cat config.new >> config
delete config.new
delete config.old
release lock

**Comment by ncw** [comment]
The traditional way of avoiding races under unix is with a hard link & rename

  * remove config.old if it exists
  * hard link config -> config.old
  * write config.new
  * rename config.new config

Those are all atomic operations and there will never be a moment when there isn't a valid config file.

However that is not cross platform and it doesn't fix the ownership problem.

Cross platform locking is a nightmare, however this appears to be a module which does the right things: [lockfile](https://github.com/nightlyone/lockfile) if we want to go in that direction.

I think I'd suggest as a compromise to use the current sequence but if rclone is running as root to preserve the ownership of the config file (or attempt to preserve it regardless).  That will mean that we can still have a nice atomic rename to fix the original bug, and it will fix this issue too.  It won't fix user1+user2 with a 777 config file, but no-one has reported that problem!

[os.Chown](https://golang.org/pkg/os/#Chown) exists for all platforms so this might actually be an easy fix.

Copying the config to config.old (or hard linking it) instead of renaming it would fix the race window.  Unfortunately Windows is fussy about renaming files over existing files.

Gah, saving config files is hard ;-)

**Comment by calisro** [comment]
If we're going to use os.chown, can we also ensure we keep the permissions of the original file as well?  At least that will ensure that my user1+user2+root use case continues to function correctly.  I believe if I give write permissions to the group both of those users are part of it should work. The ownership will bounce back and forth between the two users but it should still work.  I'd really rather not have to maintain two conf files if possible but I do understand the cross platform complexity.

**Comment by breunigs** [comment]
regarding nightlyone/lockfile: I had already a look at the code and it seems decent (so definitely worth using over rolling our own)

regarding renaming over existing files: did you check that? At least godoc claims the renaming over existing files overwrites them: https://golang.org/pkg/os/#Rename I guess I could give it a try using one of those IE11 TestDrive VMs…

**Comment by breunigs** [comment]
@calisro Here's a patched rclone version that will keep permissions and *group* only. I think that goes in more in line with the permission model's idea and should support unlimited amount of users (well, it does not include locking, so the race conditions will be a an issue at some point).

I can make it so that rclone keeps the owner the same, if it runs as root. The logic is already quite complex, so I would like to only do so if absolutely necessary. 

Logn story short, to make your use case work:
- use this version: https://www.breunig.xyz/share/2017-06-21/rclone-df293e3e884eb8cce2b58cc5d331ff9727c70f1a
- you can (must) symlink the directory the config file resides in, if you were sym/hardlinking the config file before. All rclone users must be able to read/write in that directory.
- option A: if it's okay that everyone can read/write the file, ensure you call `chown o+rw` on the config file. This should be it.
- option B: if only rclone users should be able to read/write the file, ensure every rclone user has at least one group in common. I guess `rclone` is a sensible name. Then ensure that the config file has `chown g+rw` and `chown DoesNotMatter:rlcone`, i.e. that the config file belongs to the correct group. The owner will change all the time, as you mentioned before, but the group will stay the same ensuring everyone can read/write it.

Can you give it a try and see if it works?

**Comment by breunigs** [comment]
code is in this branch: https://github.com/ncw/rclone/tree/1467-fix-config-ownership

**Comment by calisro** [comment]
I think this will work ok.  I can simply change the permissions to 660 and when root overwrites the file it will become owned by root but the group remained the original group and that original user can still read/write the file because it has group write.  I'll run it like this for a few days and see if anything weird occurs.

I did test a few scenarios where multiple mounts running (1 as the user and 4 as root) as well as adhoc rclone commands while doing the following:
-- letting it renew tokens
-- creating new remotes via 'config' as root and user
-- deleting entries as root and user

It appears to be working great.  I can live without keeping the owner consistent and rely on the group and group permissions to manage it.

Anyway, thanks a LOT for this and i'll report back if this appears to break something.  Feel free to close this issue once merged and i'll reopen if I find something broken.
